[ci] Python 3.8 macOS CI jobs failing: "symbol not found" when loading pyarrow

## Description

The Python 3.8 CI job on macOS here is failing like this:

> ___________ ERROR collecting tests/python_package_test/test_arrow.py ___________
ImportError while importing test module '/Users/runner/work/LightGBM/LightGBM/tests/python_package_test/test_arrow.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
../../../miniforge/envs/test-env/lib/python3.8/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
tests/python_package_test/test_arrow.py:21: in <module>
    import pyarrow as pa  # type: ignore
../../../miniforge/envs/test-env/lib/python3.8/site-packages/pyarrow/__init__.py:65: in <module>
    import pyarrow.lib as _lib
E   ImportError: dlopen(/Users/runner/miniforge/envs/test-env/lib/python3.8/site-packages/pyarrow/lib.cpython-38-darwin.so, 0x0002): Symbol not found: __ZN4absl12lts_2024072212log_internal10LogMessagelsIiLi0EEERS2_RKT_
E     Referenced from: <ABD8BED3-9FF3-30AE-9F2C-FB6F306CF934> /Users/runner/miniforge/envs/test-env/lib/libprotobuf.28.2.0.dylib
E     Expected in:     <E99B267F-7144-31DD-A5D4-F5B0B82894DA> /Users/runner/miniforge/envs/test-env/lib/libabsl_log_internal_message.2407.0.0.dylib
=========================== short test summary info ============================
ERROR tests/python_package_test/test_arrow.py
!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!
========================= 1 skipped, 1 error in 5.85s ==========================

([build link](https://github.com/microsoft/LightGBM/actions/runs/12585086008/job/35076093860?pr=6765))

## Reproducible example

Example build: https://github.com/microsoft/LightGBM/actions/runs/12585086008/job/35076093860?pr=6765

## Environment info

LightGBM version or commit hash:  https://github.com/microsoft/LightGBM/commit/f3bd64a1ea815f761d0f845bc16b60f0a71fdd99

## Additional Comments

I pulled the logs from a failed build ([build link](https://github.com/microsoft/LightGBM/actions/runs/12585086008/job/35076093860?pr=6765)) and the last successful one, from a few days ago ([build link](https://github.com/microsoft/LightGBM/actions/runs/12500040254/job/34875767061)).

There were just 3 changes in the conda environment:

```diff
-  libabseil          conda-forge/osx-64::libabseil-20240722.0-cxx17_hac325c4_1 
+  libabseil          conda-forge/osx-64::libabseil-20240722.0-cxx17_h0e468a2_2

-  libre2-11          conda-forge/osx-64::libre2-11-2024.07.02-hd530cb8_1 
+  libre2-11          conda-forge/osx-64::libre2-11-2024.07.02-h0e468a2_2

-  re2                conda-forge/osx-64::re2-2024.07.02-h2fb0a26_1 
+  re2                conda-forge/osx-64::re2-2024.07.02-ha5e900a_2
```

Looks like maybe this PR was the root cause:

* https://github.com/conda-forge/abseil-cpp-feedstock/pull/86

And that that led to these other discussions:

* https://github.com/conda-forge/abseil-cpp-feedstock/issues/87
* https://github.com/conda-forge/re2-feedstock/pull/87

--- a/.ci/conda-envs/ci-core-py38.txt
+++ b/.ci/conda-envs/ci-core-py38.txt
@@ -39,6 +39,9 @@ pytest=8.2.*
 # pinned here to help speed up solves
 bokeh=3.1.*
 fsspec=2024.5.*
+# TODO(jameslamb): add issue link
+libabseil=20240722.0=*_1
+libre2-11=2024.07.02=*_1
 msgpack-python=1.0.*
 pluggy=1.5.*
 pyparsing=3.1.4

--- a/.ci/conda-envs/ci-core-py38.txt
+++ b/.ci/conda-envs/ci-core-py38.txt
@@ -39,7 +39,8 @@ pytest=8.2.*
 # pinned here to help speed up solves
 bokeh=3.1.*
 fsspec=2024.5.*
-# TODO(jameslamb): add issue link
+# pinning 'libabseil' and 'libre2' to specific build numbers for pyarrow compatibility:
+# ref: https://github.com/microsoft/LightGBM/issues/6772
 libabseil=20240722.0=*_1
 libre2-11=2024.07.02=*_1
 msgpack-python=1.0.*

